docs(upload): Add KB for Antiforgery integration (#1872)
telerik/blazor-docs: knowledge-base/upload-validateantiforgerytoken.md

---
title: Upload Files with Antiforgery Validation

> description: 

Author: ```suggestion
description: Learn how to setup the Telerik Blazor Upload to work with .NET controllers that use ValidateAntiForgeryToken attribute. Configure Blazor apps with antiforgery validation.
```

type: how-to
page_title: How to Upload Files with Antiforgery Validation
slug: upload-kb-validateantiforgerytoken
position: 
tags: telerik, blazor, upload
ticketid: 1626509, 1637325
res_type: kb
---

## Environment

<table>
    <tbody>
        <tr>
            <td>Product</td>
            <td>Upload for Blazor</td>
        </tr>
    </tbody>
</table>

## Description

This KB article answers the following questions:

* How to use the Telerik Blazor Upload component with controllers that are decorated with `[ValidateAntiForgeryToken]` attribute?
* How to upload files to controllers that require antiforgery validation?
* How to set antiforgery tokens in the Upload's `OnUpload` and `OnRemove` events?

## Solution

Here are the suggested steps to configure .NET Core Blazor antiforgery validation and integrate it with the Telerik Upload component.

1. Edit `Program.cs`:

>     * Add `builder.Services.AddRazorPages();`

Reviewer: ```suggestion
    * Add `builder.Services.AddRazorPages();`.
```

>     * Add `builder.Services.AddHttpContextAccessor();`

Reviewer: ```suggestion
    * Add `builder.Services.AddHttpContextAccessor();`.
```

    * (optional) Add `builder.Services.AddAntiforgery()` with custom `HeaderName` or `FormFieldName`
    * Verify that `app.UseAntiforgery();` is present

>     * Add `app.MapDefaultControllerRoute();` to configure routing

Reviewer: ```suggestion
    * Add `app.MapDefaultControllerRoute();` to configure routing.
```

1. [Implement the `Save` and `Remove` controller methods]({%slug upload-overview%}#implement-controller-methods).

> 1. Decorate the controler class or specific action methods with `[ValidateAntiForgeryToken]`.

Reviewer: ```suggestion
1. Decorate the controller class or specific action methods with `[ValidateAntiForgeryToken]`.
```

1. Configure the Razor component, which contains the Telerik Blazor Upload:

>     * Inject `AntiforgeryStateProvider` to use its `GetAntiforgeryToken()` method

Reviewer: Consider clarifying the `its` possessive. Right now, it's not immediately clear what it stands for - is it the Razord `GetAntiforgeryToken` or the Blazor one?

```suggestion
    * Inject `AntiforgeryStateProvider` to use its `GetAntiforgeryToken()` method.
```

Author: Mentioning a dependency injection in combination with the same dependency's method or property should be intuitive to every developer. Those who get confused will rely on the code example below anyway.

>     * Inject `IAntiforgery` to use its `GetAndStoreTokens(httpContext)` method

Reviewer: Same comment about `its` as above.

```suggestion
    * Inject `IAntiforgery` to use its `GetAndStoreTokens(httpContext)` method.
```

>     * Inject `IHttpContextAccessor` to use its `HttpContext` property in the `GetAndStoreTokens()` method

Reviewer: Also, consider the general comment about `its` in these steps.

```suggestion
    * Inject `IHttpContextAccessor` to use its `HttpContext` property in the `GetAndStoreTokens()` method.
```

>     * Execute `GetAndStoreTokens()` and/or `GetAntiforgeryToken()` in `OnInitialized` to obtain the required anti-forgery information.

Reviewer: ```suggestion
    * Execute `GetAndStoreTokens()` and/or `GetAntiforgeryToken()` in `OnInitialized` to obtain the required antiforgery information.
```

>     * Add the required anti-forgery information in the Upload component's [`OnUpload` and `OnRemove` event handlers]({%slug upload-events%}).

Reviewer: ```suggestion
    * Add the required antiforgery information in the Upload component's [`OnUpload` and `OnRemove` event handlers]({%slug upload-events%}).
```

> The code snippets bellow assume that the application name is `BlazorAppName`.

Reviewer: ```suggestion
The code snippets below assume that the application name is `BlazorAppName`.
```

        if (afRequestToken != null)
        {
            AntiforgeryFormFieldName = afRequestToken.FormFieldName;
            AntiforgeryFormValue = afRequestToken.Value;
        }

        base.OnInitialized();
    }

    private string ToAbsoluteUrl(string url)
    {
        return $"{NavigationManager.BaseUri}{url}";
    }
}
````
````Controller
using Microsoft.AspNetCore.Mvc;

namespace BlazorAppName.Controllers
{
    [ValidateAntiForgeryToken]
    [Route("api/[controller]/[action]")]
    public class UploadController : ControllerBase
    {
        public IWebHostEnvironment HostingEnvironment { get; set; }

        public UploadController(IWebHostEnvironment hostingEnvironment)
        {
            HostingEnvironment = hostingEnvironment;
        }

        [HttpPost]
        public async Task<IActionResult> Save(IFormFile files)
        {
            // Save the file...

            return new EmptyResult();
        }

        [HttpPost]
        public async Task<IActionResult> Remove([FromForm] string files)
        {
            // Delete the file...

            return new EmptyResult();
        }
    }
}
````

> > This article contains code snippets and suggestions that relate to general .NET programming and antiforgery setup of a Blazor application. The provided implementation is just an example and is strictly outside the Telerik support scope. The primary resource for antiforgery configuration is the Microsoft documentation. See [Blazor authentication and authorization](https://learn.microsoft.com/en-us/aspnet/core/blazor/security/).

Reviewer: Personally, I would place this disclaimer at the top of the page or at the beginning of the Solution section.

Author: This makes sense from one point of view, but the disclaimer won't look good there, because it's dismissive (as it should be).

> > This article contains code snippets and suggestions that relate to general .NET programming and antiforgery setup of a Blazor application. The provided implementation is just an example and is strictly outside the Telerik support scope. The primary resource for antiforgery configuration is the Microsoft documentation. See [Blazor authentication and authorization](https://learn.microsoft.com/en-us/aspnet/core/blazor/security/).

Author: ```suggestion

## Disclaimer

> This article contains code snippets and suggestions that relate to general .NET programming and antiforgery setup of a Blazor application. The provided implementation is just an example and is strictly outside the Telerik support scope. The primary resource for antiforgery configuration is the Microsoft documentation. See [Blazor authentication and authorization](https://learn.microsoft.com/en-us/aspnet/core/blazor/security/).
```
